Kelpie stream10 dev bjs2 77650

JegorKuz opened the pull request.

JegorKuz committed:
[UPDATE] BJS2-77655
Частичное решение, чтобы не потерять на всякий случай.
TODO:
Настроить фильтры для поиска через jql.
Разобраться с ДТО для запроса.

JegorKuz committed:
[UPDATE] BJS2-77655
Частичное решение, чтобы не потерять на всякий случай.
TODO:
Настроить фильтры для поиска через jql.
Разобраться с ДТО для запроса.

JegorKuz committed:
[UPDATE] BJS2-77655
Частичное решение, чтобы не потерять на всякий случай.
TODO:
Настроить фильтры для поиска через jql.
Разобраться с ДТО для запроса.

JegorKuz committed:
Merge branch 'kelpie-master-stream10' into kelpie-stream10-dev-BJS2-77650

# Conflicts:
#	src/main/resources/application.yaml

JegorKuz committed:
[UPDATE] BJS2-77655
Исправил поломанный тест.

miked4king commented on src/main/java/faang/school/projectservice/config/jira/JiraConfig.java:
```diff
@@ -0,0 +1,37 @@
+    private String apiToken;
+
+    @Value("${jira.base-url}")
+    private String baseUrl;
```
если требуется несколько полей подтянуть то лучше вынести в отдельный класс, аннотировать @ConfigurationProperties и инжектить его

JegorKuz replied:
Сделал

miked4king commented on src/main/java/faang/school/projectservice/controller/jira/JiraController.java:
```diff
@@ -0,0 +1,68 @@
+    @PostMapping("/issues")
+    public ResponseEntity<?> createIssue(@RequestBody Map body) {
+        Map<String, Object> result = jiraService.createIssue(body);
+        return ResponseEntity.ok(result);
```
вопрос сразу ко всем методам контроллера:
1) почему возвращается ResponseEntity если мы не добавляем кастонмые заголовки и отдаем статус 200, который будет по умолчанию?
2) название переменной result не информативное, нужно изменить

miked4king commented on src/main/java/faang/school/projectservice/controller/jira/JiraController.java:
```diff
@@ -0,0 +1,68 @@
+                                                    @RequestBody JiraIssueFilterDto jiraTaskFilterDto,
+                                                    @RequestParam(defaultValue = "0") int startAt,
+                                                    @RequestParam(defaultValue = "100") int maxResults,
+                                                    @RequestParam(required = false) Integer limit) {
```
может лучше принять их как объект?
чтобы вместо трех полей объект в метод прокидывать

miked4king commented on src/main/java/faang/school/projectservice/controller/jira/JiraController.java:
```diff
@@ -0,0 +1,68 @@
+                                                    @RequestParam(required = false) Integer limit) {
+        if (jiraTaskFilterDto == null ||
+                (jiraTaskFilterDto.getAssignee() == null && jiraTaskFilterDto.getStatus() == null)) {
+            return getAllIssues(projectKey, startAt, maxResults, limit);
```
а почему тут вызывается метод контроллера, а не сервиса ?

miked4king commented on src/main/java/faang/school/projectservice/service/jira/JiraServiceImpl.java:
```diff
@@ -0,0 +1,230 @@
+
+    @Value("${jira.base-url}")
+    private String baseUrl;
+    private final String jiraRestApiUrl = "/rest/api/3";
```
константа?

miked4king commented on src/main/java/faang/school/projectservice/service/jira/JiraServiceImpl.java:
```diff
@@ -0,0 +1,230 @@
+    private final String jiraRestApiUrl = "/rest/api/3";
+
+    @Override
+    public Map<String, Object> createIssue(Map body) {
```
давай ответственность разделим, сделаешь отдельный компонент, который будет отправлять запросы, в данном сервисе будет происходить подготовка запроса, обработка результата и обработка ошибок

miked4king commented on src/main/java/faang/school/projectservice/service/jira/JiraServiceImpl.java:
```diff
@@ -0,0 +1,230 @@
+        Map<String, Object> allIssues = new LinkedHashMap<>();
+
+        try {
+            do {
```
почему этот запрос крутится в do while?
насколько я вижу в строках 122, 123 ты отправляешь startAt, maxResults, т.е в моем понимании ты от жиры получишь сразу нужный тебе ответ, разве не так ?

JegorKuz replied:
Мне ведь надо получить все Issues, startAt = 0, maxResults = 100 это по дефолту, их там около 80к, значит мне надо либо пройти все 80к либо до limit.
Что-бы получить все возможные, каждый новый запрос увеличивает startAt до startAt + maxResults или же startAt = limit.
Таким образом он сначала вернёт от 0 до 100 Issue потом с 100 по 200 и тд пока не вернёт все или не упрётся в лимит

JegorKuz committed:
[UPDATE] BJS2-77650
Исправил принятие тела запроса и возвращаемый ответ.

JegorKuz committed:
Merge branch 'kelpie-master-stream10' into kelpie-stream10-dev-BJS2-77650

# Conflicts:
#	src/main/resources/application.yaml

JegorKuz committed:
Merge branch 'kelpie-master-stream10' into kelpie-stream10-dev-BJS2-77650

miked4king commented on src/main/java/faang/school/projectservice/client/jira/JiraRestClient.java:
```diff
@@ -0,0 +1,39 @@
+
+    private final RestTemplate jiraRestTemplate;
+
+    public <T> ResponseEntity<T> post(String url, Object body, Class<T> bodyClass) {
```
прикольно получилось, отдаленно похоже на либу для выполнения любых запросов)
но лучше все таки сделать отдельный метод под каждый твой запрос)

JegorKuz replied:
Уж больно сильно мне понравились эти get/post/put оставил, но добавил и методы на создание каждого запроса.

miked4king commented on src/main/java/faang/school/projectservice/dto/jira/issue/request/JiraGetMultipleIssuesDto.java:
```diff
@@ -0,0 +1,12 @@
+
+@Data
+public class JiraGetMultipleIssuesDto {
+    Integer startAt;
+    Integer maxResults;
+    Integer limit;
+    JiraIssueFilterDto jiraTaskFilterDto;
```
модификаторы?

JegorKuz replied:
Ни одно из полей ведь не является обязательным?

miked4king replied:
как это влияет на модификатор доступа ?

miked4king commented on src/main/java/faang/school/projectservice/dto/jira/issue/Fields.java:
```diff
@@ -0,0 +1,121 @@
+    public static class Assignee {
+        private String id;
+    }
+    private Assignee assignee;
```
какой то странный порядок, объяви сначала поля, а потом классы, а то каша получается.
в классе JiraUpdateIssueRequest все хорошо, сделай по аналогии

JegorKuz replied:
Поправил

miked4king commented on src/main/java/faang/school/projectservice/service/jira/JiraServiceImpl.java:
```diff
@@ -0,0 +1,194 @@
+
+    private final JiraRestClient jiraRestClient;
+    private final JiraProperties jiraProperties;
+
```
лишний отступ

miked4king commented on src/main/java/faang/school/projectservice/service/jira/JiraServiceImpl.java:
```diff
@@ -0,0 +1,194 @@
+    public JiraCreateIssueResponseDto createIssue(JiraCreateIssueDto jiraCreateIssueDto) {
+        String url = new StringBuilder()
+                .append(jiraProperties.baseUrl())
+                .append(jiraProperties.restApiUrl())
```
вроде во всех методах есть jiraProperties.baseUrl() + jiraProperties.restApiUrl() - может в один параметр объединим ?

и все таки будет лучше если конструировать url мы будем в JiraRestClient, т.к он с внешней системой общается и как будто бы его зона ответственности

JegorKuz replied:
Есть контакт

miked4king commented on src/main/java/faang/school/projectservice/service/jira/JiraServiceImpl.java:
```diff
@@ -0,0 +1,194 @@
+            return response.getBody();
+        } catch (HttpClientErrorException e) {
+            log.error("Error creating task: {}", e.getResponseBodyAsString());
+            throw e;
```
тут будет классно создать кастомное исключение и бросить его, чтобы в обработчике его отловить и сделать акцент на том что проблема при вызове jira сервиса)
HttpClientErrorException - по моему 4хх ошибки, могут еще 5хх ошибки вернуться, на данный момент можешь не добавлять но на будущее держи в голове

не забудь еще добавить глобальный обработчик для обработки исключений

JegorKuz replied:
Добавил

miked4king commented on src/main/java/faang/school/projectservice/service/jira/JiraServiceImpl.java:
```diff
@@ -0,0 +1,194 @@
+                    JiraUpdateIssueResponse.class
+            );
+
+            if (response.getStatusCode().is2xxSuccessful() && response.hasBody()) {
```
response.getStatusCode().is2xxSuccessful() - проверял данный кейс? 
я имею в виду если статус будет не 200 у нас разве не будет выброшена ошибка?

JegorKuz replied:
Да, всё верно, переделал в целом логирование.
Оставил для проверки ибо не знаю как это работает с 3хх ответом, вроде не ошибка - а перенаправление, но ведь и не успешный ответ.

miked4king commented on src/main/java/faang/school/projectservice/service/jira/JiraServiceImpl.java:
```diff
@@ -0,0 +1,194 @@
+
+            return response.getBody();
+        } catch (HttpClientErrorException e) {
+            log.error("Error creating task: {}", e.getResponseBodyAsString());
```
"Error creating task: {}" - не создаем же)

miked4king commented on src/main/java/faang/school/projectservice/service/jira/JiraServiceImpl.java:
```diff
@@ -0,0 +1,194 @@
+
+            return response.getBody();
+        } catch (HttpClientErrorException e) {
+            log.error("Error creating task: {}", e.getResponseBodyAsString());
```
видимо накопипастил)

JegorKuz replied:
вынес чтоб красивше было

miked4king commented on src/main/java/faang/school/projectservice/service/jira/JiraServiceImpl.java:
```diff
@@ -0,0 +1,194 @@
+
+            return response.getBody();
+        } catch (HttpClientErrorException e) {
+            log.error("Error creating task: {}", e.getResponseBodyAsString());
```
тут тоже

miked4king reviewed (changes requested):
уже гораздо лучше, молодец

JegorKuz committed:
[UPDATE] BJS2-77650
Вынес логику создания URL(теперь вся логика запроса) в JiraRestClient.
JiraServiceImpl только обработка response и ошибок.

JegorKuz committed:
[UPDATE] BJS2-77650
Вынес логику обработки ошибок в отдельный метод.
Добавил обработку ошибок связанных с Jira.
Поправил мелкие косяки.

JegorKuz committed:
Merge remote-tracking branch 'origin/kelpie-stream10-dev-BJS2-77650' into kelpie-stream10-dev-BJS2-77650

# Conflicts:
#	src/main/resources/application.yaml

miked4king commented on src/main/java/faang/school/projectservice/client/jira/JiraRestClient.java:
```diff
@@ -11,13 +20,129 @@
 
     private final RestTemplate jiraRestTemplate;
+    private final JiraProperties jiraProperties;
+
```
для класса JiraRestClient забыли интерфейс добавить

miked4king commented on src/main/java/faang/school/projectservice/client/jira/JiraRestClient.java:
```diff
@@ -11,13 +20,129 @@
+        }
+
+        return jql.toString();
+    }
```
давай методы post, get и т.д сделаем либо приватными либо вынесем в отдельный класс чтобы ими можно было пользоваться из других клиентов.
если выберешь второе не забудь интерфейс)

JegorKuz replied:
Private, а то обидно работу терять)

miked4king commented on src/main/java/faang/school/projectservice/service/jira/JiraServiceImpl.java:
```diff
@@ -1,194 +1,187 @@
-            jql.append("+AND+")
-                    .append("assignee=")
-                    .append(jiraIssueFilterDto.getAssignee());
+    private void logSuccessfulResponse(ResponseEntity<?> response) {
+        if (response.getStatusCode().is2xxSuccessful()) {
+            log.info("Successful response with code: {}", response.getStatusCode());
+            log.info("Response body:\n{}", response.getBody());
```
в целом можно и не логгировать успешный ответ, у тебя же там огромные dto))
а если у нас будет множество запросов в логах будет трэш

JegorKuz replied:
Согласен

miked4king reviewed (changes requested):
давай последние правки сделаем и будем заканчивать с этой задачей)

JegorKuz committed:
[UPDATE] BJS2-77650
Ну вроде закончили)
